pkg/controller/common.go

```diff
@@ -158,103 +161,123 @@ func (c *SharedState) getPodsFromVolume(volumeName string) ([]string, error) {
 				if primaryNode == "" {
 					primaryNode = azVolumeAttachment.Spec.NodeName
 				}
-				continue
-			}
-			if _, ok := nodeScores[azVolumeAttachment.Spec.NodeName]; !ok {
-				nodeScores[azVolumeAttachment.Spec.NodeName] = 0
-				nodes = append(nodes, azVolumeAttachment.Spec.NodeName)
+			} else {
+				klog.V(5).Info(fmt.Sprintf("Azvolumeattachment %s is not primary. Scoring node %s with replica attachment. Current score array is %v. Len: %d.", azVolumeAttachment.Name, azVolumeAttachment.Spec.NodeName, nodeScores, len(nodeScores)))
+				if _, ok := nodeScores[azVolumeAttachment.Spec.NodeName]; !ok {
+					// Node is not in the arrayl. Adding node to the array with score 0.
+					nodeScores[azVolumeAttachment.Spec.NodeName] = 0
+					nodes = append(nodes, azVolumeAttachment.Spec.NodeName)
+				}
+				nodeScores[azVolumeAttachment.Spec.NodeName]++
+				klog.V(5).Info(fmt.Sprintf("New score for node %s is %d. Updated score array is %v. Len: %d.", azVolumeAttachment.Spec.NodeName, nodeScores[azVolumeAttachment.Spec.NodeName], nodeScores, len(nodeScores)))
 			}
-			nodeScores[azVolumeAttachment.Spec.NodeName]++
 		}
 	}
+	// sorting nodes array per their score in nodeScore array
 	sort.Slice(nodes[:], func(i, j int) bool {
 		return nodeScores[nodes[i]] > nodeScores[nodes[j]]
 	})
 
+	// select at least as many nodes as requested replicas
 	if len(nodes) < numReplica {
 		selected, err := selectNodes(ctx, azr, numReplica-len(nodes), primaryNode, nodeScores)
 		if err != nil {
 			return nil, err
 		}
 		nodes = append(nodes, selected...)
 	}
-	return nodes[:min(len(nodes), numReplica)], nil
+	return nodes[:numReplica], nil
 }
 
-func (c *SharedState) addPod(pod *corev1.Pod, updateOption updateWithLock) (podLock *sync.Mutex) {
-	podKey := getQualifiedName(pod.Namespace, pod.Name)
+func (c *SharedState) addPod(pod *v1.Pod, updateOption updateWithLock) {
 
+	podKey := getUniqueNamespacedName(pod.Namespace, pod.Name)
+	klog.V(5).Info(fmt.Sprintf("Adding pod %s to shared map with keyName %s.", pod.Name, podKey))
 	v, _ := c.podLocks.LoadOrStore(podKey, &sync.Mutex{})
-	podLock = v.(*sync.Mutex)
+
+	podLock := v.(*sync.Mutex)
 	if updateOption == acquireLock {
 		podLock.Lock()
+		defer podLock.Unlock()
 	}
-
+	klog.V(5).Info(fmt.Sprintf("Pod spec of pod %s is: %v. With volumes: %v", pod.Name, pod.Spec, pod.Spec.Volumes))
 	claims := []string{}
 	for _, volume := range pod.Spec.Volumes {
-		if volume.CSI == nil || volume.CSI.Driver != azureutils.DriverName {
+
+		if volume.PersistentVolumeClaim == nil {
+			klog.V(5).Info(fmt.Sprintf("Pod %s: Skipping Volume %s. No persistent volume exists.", pod.Name, volume))
 			continue
 		}
-		namespacedClaimName := getQualifiedName(pod.Namespace, volume.PersistentVolumeClaim.ClaimName)
+		namespacedClaimName := getUniqueNamespacedName(pod.Namespace, volume.PersistentVolumeClaim.ClaimName)
+		_, ok := c.claimToVolumeMap.Load(namespacedClaimName)
```

Comment:
consider `if _, ok := c.claimToVolumeMap.Load(namespacedClaimName); !ok { ... }` to keep the style consistent.